### PR TITLE
[v6-exp] Consolidate in-flight changes

### LINKED ISSUE #1155: Symlinks pointing outside repository are mangled with PlainOpen
See here for fix: https://github.com/go-git/go-git/issues/253#issuecomment-2043338952

I need to commit symlinks that are intentionally pointed outside the repository.

For example:

```
# git diff X11/fontpath.d/liberation-sans-fonts
diff --git a/X11/fontpath.d/liberation-sans-fonts b/X11/fontpath.d/liberation-sans-fonts
index 0e445f8..0a9a11f 120000
--- a/X11/fontpath.d/liberation-sans-fonts
+++ b/X11/fontpath.d/liberation-sans-fonts
@@ -1 +1 @@
-/../usr/share/fonts/liberation-sans
\ No newline at end of file
+/usr/share/fonts/liberation-sans
\ No newline at end of file
# ls -l X11/fontpath.d/liberation-sans-fonts
lrwxrwxrwx. 1 root root 32 Apr  1 20:59 X11/fontpath.d/liberation-sans-fonts -> /usr/share/fonts/liberation-sans
```

`git` properly commits this as-is, but go-git mangles it to an absolute path in the commit - THUS even after a commit, the worktree is always dirty *because the commit created by go-git does not contain the proper content that reflects the real 'contents'*.

This leads to broken status in *addition* to an inaccurate/incorrect commit's file contents.

### PR DESCRIPTION
**Pull request by pjbgf**
A few changes aimed at the `v6-exp` branch were floating around different branches. This PR consolidates those changes so that those other branches can be deleted.

Changes of note:
- 4a05b0beaeb9e96da3dd0405ac93e39d5983fb41 Enable read-concurrency for commits and tags
- 299d3694fe8bf1b13818022899d0dd78acc08b35 v2 protocol enablement
- dc3adbb94e1ca7f9cb904bf1a2743e2d74a08ffb Make `osfs.BoundOS` the default `osfs` (fixes #1155).

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -109,6 +116,20 @@ type Config struct {
 		ObjectFormat format.ObjectFormat
 	}
 
+	Protocol struct {
+		// Version sets the preferred version for the Git wire protocol.
+		// When set, clients will attempt to communicate with a server
+		// using the specified protocol version. If the server does not
+		// support it, communication falls back to version 0. If unset,
+		// the default is 2. Supported versions:
```
Comment (aymanbagabas) [inline config/config.go:124]:
```suggestion
		// using the specified protocol version. If the server does not
		// support it, communication falls back to version 0. If unset,
		// the default version will be used. Supported versions:
```

Comment (aymanbagabas) [reply config/config.go:124]: Support depends on the implementation of both client and server.

File: config/config.go [hunk 2]
```diff
@@ -15,13 +15,20 @@ import (
...
 	// DefaultFetchRefSpec is the default refspec used for fetch.
 	DefaultFetchRefSpec = "+refs/heads/*:refs/remotes/%s/*"
 	// DefaultPushRefSpec is the default refspec used for push.
 	DefaultPushRefSpec = "refs/heads/*:refs/heads/*"
+	// DefaultProtocolVersion is the value assumed if none is defined
+	// at the config file. This value is used to define when this section
+	// should be marshalled or not.
+	// Note that this does not need to align with the default protocol
+	// version from plumbing/protocol.
+	DefaultProtocolVersion = protocol.Unknown
```
Comment (aymanbagabas) [inline config/config.go:31]:
```suggestion
	DefaultProtocolVersion = protocol.V0 // go-git only supports V0 for the moment
```

File: config/config.go [hunk 3]
```diff
@@ -109,6 +116,20 @@ type Config struct {
...
+		// Version sets the preferred version for the Git wire protocol.
+		// When set, clients will attempt to communicate with a server
+		// using the specified protocol version. If the server does not
+		// support it, communication falls back to version 0. If unset,
+		// the default is 2. Supported versions:
+		//
+		//   0 - the original wire protocol.
+		//   1 - the original wire protocol with the addition of a
+		// version string in the initial response from the server.
+		//   2 - Wire protocol version 2.
```
Comment (aymanbagabas) [inline config/config.go:129]: Fwiw, go-git only supports V0 as of now

File: plumbing/protocol/version.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	"fmt"
+)
+
+var ErrUnknownProtocol = errors.New("unknown Git Wire protocol")
+
+// Version sets the preferred version for the Git wire protocol.
+type Version int
+
+const (
+	Unknown Version = -1
```
Comment (aymanbagabas) [inline plumbing/protocol/version.go:14]:
```suggestion
	Undefined Version = -1
```
Perhaps `Undefined` signifies the intention better

File: _examples/custom_http/main.go [hunk 5]
```diff
@@ -36,7 +36,7 @@ func main() {
 	}
 
 	// Override http(s) default protocol to use our custom client
-	client.InstallProtocol("https", githttp.NewClient(customClient))
+	transport.Register("https", githttp.NewClient(customClient))
```
Comment (aymanbagabas) [inline _examples/custom_http/main.go:39]: :chefkiss:

**Review by aymanbagabas — COMMENTED** [review]
LGTM, some minor design comments

**Comment by aymanbagabas** [comment]
@pjbgf Thinking more about `protocol.Undefined`, maybe that's unnecessary. Is there a way to get rid of that and make it so that it uses the config defined version, or falls back to the `DefaultProtocolVersion`?

**Comment by pjbgf** [comment]
@aymanbagabas good point. I think we are better positioned to review the mechanics around this during the implementation of newer protocol versions, so we can test it E2E. I am aiming to work on this soonish - just trying to get other minor changes on v6-exp out of the way first.

**Pull request merged**